src/cfn-custom-resources/client-secret-retrieval/index.ts

```diff
@@ -0,0 +1,75 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+
+import {
+    CloudFormationCustomResourceHandler,
+    CloudFormationCustomResourceResponse,
+    CloudFormationCustomResourceDeleteEvent,
+    CloudFormationCustomResourceUpdateEvent
+} from 'aws-lambda';
+import axios from 'axios';
+import CognitoIdentityServiceProvider from 'aws-sdk/clients/cognitoidentityserviceprovider';
+
+const COGNITO_CLIENT = new CognitoIdentityServiceProvider({ region: process.env.AWS_REGION || process.env.AWS_DEFAULT_REGION });
+
+async function retrieveClientSecret(
+    action: 'Create' | 'Update' | 'Delete',
+    userPoolId: string,
+    clientId: string,
+    physicalResourceId?: string) {
+    if (action === 'Delete') {
+        // Deletes aren't executed; the standard Resource should just be deleted
+        return { physicalResourceId: physicalResourceId };
+    }
+    const input: CognitoIdentityServiceProvider.Types.DescribeUserPoolClientRequest = {
+        UserPoolId: userPoolId,
+        ClientId: clientId,
+    };
+    const res: CognitoIdentityServiceProvider.Types.DescribeUserPoolClientResponse =
+        await COGNITO_CLIENT.describeUserPoolClient(input).promise();
+    return {
+        physicalResourceId: `${userPoolId}-${clientId}-retrieved-client-secret`,
+        Data: {
+            ClientSecret: (res.UserPoolClient && res.UserPoolClient.ClientSecret) ? res.UserPoolClient.ClientSecret : '',
```

Comment:
Propose to just go at it like:

```
res.UserPoolClient!.ClientSecret
```

It reads nicer than the ternary (maybe an opinion), and you know it will have a value because you only instantiate this Custom Resource when you want to extract the client secret

Reply:
Updated